glean-core/python/glean/_ffi.py

```diff
@@ -96,19 +92,44 @@ def ffi_encode_vec_int32(values: List[int]) -> Any:
     return ffi.new("int32_t []", values)
 
 
-def ffi_decode_string(cdata) -> str:
+def ffi_decode_string(cdata, free_memory=True) -> str:
     """
-    Convert a string returned from Rust to a Python string, and free the Rust
+    Convert a string returned from Rust to a Python string, and optionally free the Rust
     string.
+
+    Args:
+        cdata: The C data containing the string value.
+        free_memory (bool): Whether or not to free the memory allocated in C.
     """
     try:
         return ffi.string(cdata).decode("utf-8")
     finally:
-        lib.glean_str_free(cdata)
+        if free_memory:
+            lib.glean_str_free(cdata)
+
+
+def ffi_decode_byte_buffer(byte_buffer, free_memory=True) -> bytes:
+    """
+    Convert a ByteBuffer returned from Rust to a Python bytes object, and optionally
+    free the Rust buffer.
+
+    Args:
+        byte_buffer: The C byte buffer.
+        free_memory (bool): Whether or not to free the memory allocated in C.
+    """
+    try:
+        # This copies the buffer, which is slower, but makes memory management
+        # easier because we don't have to remember to free the Rust side with
+        # glean_str_free() later.
+        return bytes(ffi.buffer(byte_buffer.data, byte_buffer.len))
```

Comment:
@mdboom where does the copy actually happen, here? With `bytes`?

Reply:
Yes.  `ffi.buffer` returns a Python `buffer` object, which is basically just a thin wrapper around the pointer and length on the Rust side.  `bytes` copies that in a new object whose memory is managed by Python.  The question to confirm is -- is the Python caller responsible for freeing the memory passed over in the BytesBuffer object?  If so, this fix (with the copy) is probably the easiest way to reliably handle that.  If we don't, we could remove the `bytes` and the `free` here.

Reply:
Nope, freeing memory is glean-core responsibility. I will fix this.